File: leetcode/c++/medium/_2657_find_the_prefix_common_array_of_two_arrays.cpp
// Leetcode Problem 2657

#include <gtest/gtest.h>

using namespace std;

class Solution {
public:
    vector<int> findThePrefixCommonArray(vector<int>& A, vector<int>& B) {
        int n = A.size();
        vector<int> frequency(n + 1, 0);
        vector<int> prefixCommonArray(n);

        int common = 0;
        for (int i = 0; i < n; i++) {
            if (++frequency[A[i]] == 2) common++;
            if (++frequency[B[i]] == 2) common++;
            prefixCommonArray[i] = common;
        }

        return prefixCommonArray;
    }
};


class TestFindPrefixCommonArray : public ::testing::Test {
protected:
    Solution solution;
};

TEST_F(TestFindPrefixCommonArray, SimpleTestCase1) {
    vector<int> A = {1, 3, 2, 4};
    vector<int> B = {3, 1, 2, 4};
    vector<int> expected = {0, 2, 3, 4};

    EXPECT_EQ(solution.findThePrefixCommonArray(A, B), expected);
}

TEST_F(TestFindPrefixCommonArray, SimpleTestCase2) {
    vector<int> A = {2, 3, 1};
    vector<int> B = {3, 1, 2};
    vector<int> expected = {0, 1, 3};

    EXPECT_EQ(solution.findThePrefixCommonArray(A, B), expected);
}

File: leetcode/rust/problems/src/medium/_2657_find_the_prefix_common_array_of_two_arrays.rs
// Leetcode Problem 2657

struct Solution;

#[allow(dead_code)]
impl Solution {
    pub fn find_the_prefix_common_array(a: Vec<i32>, b: Vec<i32>) -> Vec<i32> {
        let n: usize = a.len();
        let mut frequency: Vec<i32> = vec![0; n+1];
        let mut solution: Vec<i32> = vec![0; n];

        let mut common: i32 = 0;
        for i in 0..n {
            if { frequency[a[i] as usize] += 1; frequency[a[i] as usize] } == 2 {
                common += 1;
            }

            if { frequency[b[i] as usize] += 1; frequency[b[i] as usize] } == 2 {
                common += 1;
            }

            solution[i] = common;
        }

        solution
    }
}


#[test]
fn test_find_the_prefix_common_array() {
    let a1 = vec![1, 3, 2, 4];
    let b1 = vec![3, 1, 2, 4];
    let expected1 = vec![0, 2, 3, 4];

    let a2 = vec![2, 3, 1];
    let b2 = vec![3, 1, 2];
    let expected2 = vec![0, 1, 3];

    assert_eq!(Solution::find_the_prefix_common_array(a1, b1), expected1);
    assert_eq!(Solution::find_the_prefix_common_array(a2, b2), expected2);
}